### ISSUE TITLE
collation function error

### ISSUE THREAD

**Issue by jingshanglu** [pingcap/tidb#28150]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
DROP TABLE IF EXISTS t1;
CREATE TABLE t1 ( a int, p1 VARCHAR(255) CHARACTER SET utf8 COLLATE utf8_bin,p2 VARCHAR(255) CHARACTER SET utf8 COLLATE utf8_general_ci , p3 VARCHAR(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_bin,p4 VARCHAR(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci ,n1 VARCHAR(255) CHARACTER SET utf8 COLLATE utf8_bin,n2 VARCHAR(255) CHARACTER SET utf8 COLLATE utf8_general_ci , n3 VARCHAR(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_bin,n4 VARCHAR(255) CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci );

insert into t1 (a) values (1);
insert into t1 (a,p1,p2,p3,p4,n1,n2,n3,n4) values (2,'0aA1!测试テストמבחן ','0aA1!测试テストמבחן ','0aA1!测试テストמבחן ','0aA1!测试テストמבחן ','0Aa1!测试テストמבחן','0Aa1!测试テストמבחן','0Aa1!测试テストמבחן','0Aa1!测试テストמבחן');
insert into t1 (a,p1,p2,p3,p4,n1,n2,n3,n4) values (3,'0aA1!测试テストמבחן','0aA1!测试テストמבחן','0aA1!测试テストמבחן','0aA1!测试テストמבחן','0Aa1!测试テストמבחן','0Aa1!测试テストמבחן','0Aa1!测试テストמבחן','0Aa1!测试テストמבחן');
insert into t1 (a,p1,p2,p3,p4,n1,n2,n3,n4) values (4,'0aA1!测试テストמבחן','0aA1!测试テストמבחן','0aA1!测试テストמבחן','0aA1!测试テストמבחן','0Aa1!测试テストמבחן ','0Aa1!测试テストמבחן ','0Aa1!测试テストמבחן ','0Aa1!测试テストמבחן ');
insert into t1 (a,p1,p2,p3,p4,n1,n2,n3,n4) values (5,'0aA1!测试テストמבחן0aA1!测试','0aA1!测试テストמבחן0aA1!测试','0aA1!测试テストמבחן0aA1!测试','0aA1!测试テストמבחן0aA1!测试','0Aa1!测试','0Aa1!测试','0Aa1!测试','0Aa1!测试');
insert into t1 (a,p1,p2,p3,p4,n1,n2,n3,n4) values (6,'0aA1!测试テストמבחן0aA1!测试','0aA1!测试テストמבחן0aA1!测试','0aA1!测试テストמבחן0aA1!测试','0aA1!测试テストמבחן0aA1!测试','0aA1!测试','0aA1!测试','0aA1!测试','0aA1!测试');
insert into t1 (a,p1,p2,p3,p4,n1,n2,n3,n4) values (7,'0aA1!测试テストמבחן ','0aA1!测试テストמבחן ','0aA1!测试テストמבחן ','0aA1!测试テストמבחן ','0aA1!测试テストמבחן','0aA1!测试テストמבחן','0aA1!测试テストמבחן','0aA1!测试テストמבחן');
insert into t1 (a,p1,p2,p3,p4,n1,n2,n3,n4) values (8,'0aA1!测试テストמבחן','0aA1!测试テストמבחן','0aA1!测试テストמבחן','0aA1!测试テストמבחן','0aA1!测试テストמבחן ','0aA1!测试テストמבחן ','0aA1!测试テストמבחן ','0aA1!测试テストמבחן ');

select cast(p1 as char character set utf8mb4) from t1;
select cast(p2 as char character set utf8mb4) from t1;
select cast(p3 as char character set utf8mb4) from t1;
select cast(p4 as char character set utf8mb4) from t1;
select cast(p1 as char character set utf8) from t1;
select cast(p2 as char character set utf8) from t1;
select cast(p3 as char character set utf8) from t1;
select cast(p4 as char character set utf8) from t1;

select collation(p1) = collation(cast(p1 as char)) from t1;
select collation(p2) = collation(cast(p2 as char)) from t1;
select collation(p3) = collation(cast(p3 as char)) from t1;
select collation(p4) = collation(cast(p4 as char)) from t1;
```

ticase: https://internal.pingcap.net/jira/browse/TICASE-1692, https://internal.pingcap.net/jira/browse/TICASE-1693

### 2. What did you expect to see? (Required)

```
mysql> select collation(p3) = collation(cast(p3 as char)) from t1;
+---------------------------------------------+
| collation(p3) = collation(cast(p3 as char)) |
+---------------------------------------------+
|                                           0 |
|                                           0 |
|                                           0 |
|                                           0 |
|                                           0 |
|                                           0 |
|                                           0 |
|                                           0 |
+---------------------------------------------+
```

### 3. What did you see instead (Required)

```
mysql> select collation(p3) = collation(cast(p3 as char)) from t1;
+---------------------------------------------+
| collation(p3) = collation(cast(p3 as char)) |
+---------------------------------------------+
|                                           1 |
|                                           1 |
|                                           1 |
|                                           1 |
|                                           1 |
|                                           1 |
|                                           1 |
|                                           1 |
+---------------------------------------------+
```

### 4. What is your TiDB version? (Required)

4.0.14

**Comment by xiongjiwei** [comment]
this is a bug in tidb 4.0, which has been fixed in tidb 5.0

**Comment by bb7133** [comment]
Close this issue since it has been fixed in the latest version.
